### PR TITLE
Tpetra: Fix #4797 & fix #4789

### LINKED ISSUE #4789: Unused Tpetra var warning in header breaking SPARC Trilinos Integration builds with -Werror
CC: @trilinos/tpetra, @kddevin  (Trilinos Data Services Product Lead), @bartlettroscoe, @fryeguy52, @mhoemmen 

## Next Action Status

<status-and-or-first-action>

## Description

Recently Tpetra was updated to add a new unused var warning:

```
.../Tpetra_Details_StaticView.hpp:168:9: error: unused type alias 'host_device_type' [-Werror,-Wunused-local-typedef]
```

This is bringing down the [SPARC Trilinos Integration builds](https://snl-wiki.sandia.gov/display/CoodinatedDevOpsATDM/SPARC+Trilinos+Integration+Process) for Clang and GCC due to the SPARC setting `-Werror` in those builds.  Therefore, we get no build or test results in those builds due to this.

## Current Status on CDash

The current status can only be seen in the SPARC Trilinos Integration builds shown over the last few days [here](http://compsim-dashboard.sandia.gov/cdash/index.php?project=SPARC&date=2019-03-30&filtercombine=and&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=66&value1=-trildev&field2=buildstarttime&compare2=83&value2=5%20days%20ago).

## Steps to Reproduce

Instructions to reproduce builds are at:

* https://github.com/trilinos/Trilinos/blob/develop/cmake/std/atdm/README.md

More specifically, the commands given for the system 'cee-rhel6' are provided at:

* https://github.com/trilinos/Trilinos/blob/develop/cmake/std/atdm/README.md#cee-rhel6-environment

The exact commands to reproduce this warning should (hopefully) be:

```
$ cd <some_build_dir>/

$ source $TRILINOS_DIR/cmake/std/atdm/load-env.sh \
    Trilinos-atdm-cee-rhel6_clang-5.0.1_openmpi-1.10.2_serial_static_opt

$ cmake \
 -GNinja \
 -DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/atdm/ATDMDevEnv.cmake \
 -DTrilinos_ENABLE_TESTS=ON -DTrilinos_ENABLE_Tpetra=ON \
 $TRILINOS_DIR

$ make NP=16
```

### LINKED ISSUE #4797: Tpetra::MultiVector: Deprecate getDualView & remove its use from Trilinos
@trilinos/tpetra 

- [ ] Deprecate `Tpetra::MultiVector::getDualView`.
- [ ] Remove its use from Trilinos.
- [ ] Explain (in the method's Doxygen documentation) how to get the same functionality.

## Motivation and Context

The `getDualView` method exposes an implementation detail of `Tpetra::MultiVector`.  This hinders fixing #364 and #333.

### PR DESCRIPTION
**Pull request by mhoemmen**
@trilinos/tpetra 

This supersedes PR #4798.

## Description

  - Deprecate `Tpetra::MultiVector::getDualView` method.
  - Add new `Tpetra::Vector` "offset view" constructor so that downstream code doesn't need this method any more.  
  - Remove all use of this method from downstream Trilinos packages.
  - Document how to get the same functionality elsewhere.
  - Fix unrelated build warnings.

## Motivation and Context

The deprecated method exposes an implementation detail of `Tpetra::MultiVector` and `Tpetra::Vector`, namely that they have a `Kokkos::DualView` inside.  This hinders fixing outstanding issues like #364, and also hinders optimizations like #333.

## Related Issues

* Closes #4797, #4789 
* Related to #333, #364 

## How Has This Been Tested?

Fully downstream with GCC + OpenMP.

### CODE REVIEW

File: packages/tpetra/core/src/Tpetra_CrsMatrix_def.hpp [hunk 1]
```diff
@@ -4273,33 +4271,35 @@ namespace Tpetra {
...
-    diag_dv.clear_sync_state();
+    diag.clear_sync_state ();
 
     // For now, we fill the Vector on the host and sync to device.
     // Later, we may write a parallel kernel that works entirely on
     // device.
-    diag.modify_host();
-    auto lclVecHost = diag.template getLocalView<host_execution_space> ();
+    diag.modify_host ();
+    auto lclVecHost = diag.getLocalViewHost ();
```
Comment (lucbv) [inline packages/tpetra/core/src/Tpetra_CrsMatrix_def.hpp]: Nice I have dislike that `template` qualifier in the middle of the method call for a while : )

File: packages/tpetra/core/src/Tpetra_Details_crsUtils.hpp [hunk 2]
```diff
@@ -170,33 +171,40 @@ insert_crs_indices(
...
 
-  using ordinal = typename InOutIndices::value_type;
-  auto const start = row_ptrs[row];
-  auto end = start + num_assigned;
-  auto num_avail = row_ptrs[row + 1] - end;
+  const offset_type start = row_ptrs[row];
+  offset_type end = start + static_cast<offset_type> (num_assigned);
+  const size_t num_avail = (row_ptrs[row + 1] < end) ? size_t (0) :
+    row_ptrs[row + 1] - end;
+  const size_t num_new_indices = static_cast<size_t> (new_indices.size ());
```
Comment (lucbv) [inline packages/tpetra/core/src/Tpetra_Details_crsUtils.hpp]: Is not new_indices.size() already a `size_t`?

File: packages/tpetra/core/src/Tpetra_Details_crsUtils.hpp [hunk 3]
```diff
@@ -170,33 +171,40 @@ insert_crs_indices(
...
-  {
-    auto row_offset = start;
-    ordinal idx = std::forward<IndexMap>(map)(new_indices[k]);
-    for (; row_offset < end; row_offset++)
-      if (idx == cur_indices[row_offset])
+
+  for (size_t k = 0; k < num_new_indices; ++k) {
+    const ordinal_type idx = std::forward<IndexMap>(map)(new_indices[k]);
+    offset_type row_offset = start;
+    for (; row_offset < end; ++row_offset) {
```
Comment (lucbv) [inline packages/tpetra/core/src/Tpetra_Details_crsUtils.hpp]: Any reason to declare the iterated variable outside of the for statement?

File: packages/tpetra/core/src/Tpetra_Details_crsUtils.hpp (hunk 2 above)
Comment (mhoemmen) [reply packages/tpetra/core/src/Tpetra_Details_crsUtils.hpp]: This code is super generic and should work for `std::vector`, `Teuchos::Array*`, or `Kokkos::View`.  `Teuchos::Array*<T>::size_type` is `ptrdiff_t`, not `size_t`.

File: packages/tpetra/core/src/Tpetra_Details_crsUtils.hpp (hunk 3 above)
Comment (mhoemmen) [reply packages/tpetra/core/src/Tpetra_Details_crsUtils.hpp]: @lucbv The code uses that variable after the for loop.  It's like `std::find` returning the end iterator.  @tjfulle wrote this code if you have more questions :-)

**Comment by mhoemmen** [comment]
@lucbv Thanks for the review! :D

**Pull request merged**